src/query/ast/src/parser/statement.rs

```diff
@@ -1926,13 +1926,46 @@ pub fn grant_source(i: Input) -> IResult<AccountMgrSource> {
         },
     );
     let all = map(
-        rule! { ALL ~ PRIVILEGES? ~ ON ~ #grant_level },
+        rule! { ALL ~ PRIVILEGES? ~ ON ~ #grant_all_level },
         |(_, _, _, level)| AccountMgrSource::ALL { level },
     );
 
+    let udf_privs = map(
+        rule! {
+            USAGEUDF ~ ON ~ UDF ~ #ident
+        },
+        |(_, _, _, udf)| AccountMgrSource::Privs {
+            privileges: vec![UserPrivilegeType::UsageUDF],
+            level: AccountMgrLevel::UDF(udf.to_string()),
+        },
+    );
+
+    let udf_all_privs = map(
+        rule! {
+            ALL ~ PRIVILEGES? ~ ON ~ UDF ~ #ident
+        },
+        |(_, _, _, _, udf)| AccountMgrSource::Privs {
+            privileges: vec![UserPrivilegeType::UsageUDF],
+            level: AccountMgrLevel::UDF(udf.to_string()),
+        },
+    );
+
+    let stage_privs = map(
+        rule! {
+            #comma_separated_list1(stage_priv_type) ~ ON ~ STAGE ~ #ident
```

Comment:
since we have a concrete stage name, we should already know it`s type, why does user have to provide stage_priv_type here?